File: internal/server/usage.go
package server

import (
	"go.kenn.io/agentsview/internal/db"
	"go.kenn.io/agentsview/internal/export"
	"go.kenn.io/agentsview/internal/service"
)

// Comparison holds the prior-period cost comparison returned by
// GET /api/v1/usage/comparison.
type Comparison struct {
	PriorFrom      string  `json:"priorFrom"`
	PriorTo        string  `json:"priorTo"`
	PriorTotalCost float64 `json:"priorTotalCost"`
	DeltaPct       float64 `json:"deltaPct"`
}

// ProjectTotal holds range-wide token and cost totals per project.
type ProjectTotal struct {
	Project             string  `json:"project"`
	InputTokens         int     `json:"inputTokens"`
	OutputTokens        int     `json:"outputTokens"`
	CacheCreationTokens int     `json:"cacheCreationTokens"`
	CacheReadTokens     int     `json:"cacheReadTokens"`
	Cost                float64 `json:"cost"`
}

// ModelTotal holds range-wide token and cost totals per model.
type ModelTotal struct {
	Model               string  `json:"model"`
	InputTokens         int     `json:"inputTokens"`
	OutputTokens        int     `json:"outputTokens"`
	CacheCreationTokens int     `json:"cacheCreationTokens"`
	CacheReadTokens     int     `json:"cacheReadTokens"`
	Cost                float64 `json:"cost"`
}

// AgentTotal holds range-wide token and cost totals per agent.
type AgentTotal struct {
	Agent               string  `json:"agent"`
	InputTokens         int     `json:"inputTokens"`
	OutputTokens        int     `json:"outputTokens"`
	CacheCreationTokens int     `json:"cacheCreationTokens"`
	CacheReadTokens     int     `json:"cacheReadTokens"`
	Cost                float64 `json:"cost"`
}

// BranchTotal holds range-wide token and cost totals per (project, branch).
type BranchTotal struct {
	Project             string  `json:"project"`
	Branch              string  `json:"branch"`
	InputTokens         int     `json:"inputTokens"`
	OutputTokens        int     `json:"outputTokens"`
	CacheCreationTokens int     `json:"cacheCreationTokens"`
	CacheReadTokens     int     `json:"cacheReadTokens"`
	Cost                float64 `json:"cost"`
}

// CacheStats summarizes cache hit/miss for the period.
type CacheStats struct {
	CacheReadTokens     int     `json:"cacheReadTokens"`
	CacheCreationTokens int     `json:"cacheCreationTokens"`
	UncachedInputTokens int     `json:"uncachedInputTokens"`
	OutputTokens        int     `json:"outputTokens"`
	HitRate             float64 `json:"hitRate"`
	SavingsVsUncached   float64 `json:"savingsVsUncached"`
}

type UnsupportedUsage struct {
	Kind string `json:"kind"`
}

// UsageSummaryResponse preserves the public OpenAPI schema for
// GET /api/v1/usage/summary while the handler delegates assembly to
// the transport-neutral service layer.
type UsageSummaryResponse struct {
	SchemaVersion    int                               `json:"schema_version,omitempty"`
	Pricing          *export.PricingBlock              `json:"pricing,omitempty"`
	Projects         map[string]export.ProjectMapEntry `json:"projects"`
	From             string                            `json:"from"`
	To               string                            `json:"to"`
	Totals           db.UsageTotals                    `json:"totals"`
	Daily            []db.DailyUsageEntry              `json:"daily"`
	ProjectTotals    []ProjectTotal                    `json:"projectTotals"`
	ModelTotals      []ModelTotal                      `json:"modelTotals"`
	AgentTotals      []AgentTotal                      `json:"agentTotals"`
	BranchTotals     []BranchTotal                     `json:"branchTotals"`
	SessionCounts    db.UsageSessionCounts             `json:"sessionCounts"`
	CacheStats       CacheStats                        `json:"cacheStats"`
	UnsupportedUsage *UnsupportedUsage                 `json:"unsupportedUsage,omitempty"`
	Comparison       *Comparison                       `json:"comparison,omitempty"`
}

func usageSummaryResponseFromService(
	res *service.UsageSummaryResult,
) UsageSummaryResponse {
	return UsageSummaryResponse{
		SchemaVersion:    res.SchemaVersion,
		Pricing:          res.Pricing,
		Projects:         res.Projects,
		From:             res.From,
		To:               res.To,
		Totals:           res.Totals,
		Daily:            res.Daily,
		ProjectTotals:    projectTotalsFromService(res.ProjectTotals),
		ModelTotals:      modelTotalsFromService(res.ModelTotals),
		AgentTotals:      agentTotalsFromService(res.AgentTotals),
		BranchTotals:     branchTotalsFromService(res.BranchTotals),
		SessionCounts:    res.SessionCounts,
		CacheStats:       cacheStatsFromService(res.CacheStats),
		UnsupportedUsage: unsupportedUsageFromService(res.UnsupportedUsage),
	}
}

func unsupportedUsageFromService(
	in *service.UnsupportedUsage,
) *UnsupportedUsage {
	if in == nil {
		return nil
	}
	return &UnsupportedUsage{Kind: in.Kind}
}

func projectTotalsFromService(in []service.ProjectTotal) []ProjectTotal {
	out := make([]ProjectTotal, 0, len(in))
	for _, total := range in {
		out = append(out, ProjectTotal{
			Project:             total.Project,
			InputTokens:         total.InputTokens,
			OutputTokens:        total.OutputTokens,
			CacheCreationTokens: total.CacheCreationTokens,
			CacheReadTokens:     total.CacheReadTokens,
			Cost:                total.Cost,
		})
	}
	return out
}

func modelTotalsFromService(in []service.ModelTotal) []ModelTotal {
	out := make([]ModelTotal, 0, len(in))
	for _, total := range in {
		out = append(out, ModelTotal{
			Model:               total.Model,
			InputTokens:         total.InputTokens,
			OutputTokens:        total.OutputTokens,
			CacheCreationTokens: total.CacheCreationTokens,
			CacheReadTokens:     total.CacheReadTokens,
			Cost:                total.Cost,
		})
	}
	return out
}

func agentTotalsFromService(in []service.AgentTotal) []AgentTotal {
	out := make([]AgentTotal, 0, len(in))
	for _, total := range in {
		out = append(out, AgentTotal{
			Agent:               total.Agent,
			InputTokens:         total.InputTokens,
			OutputTokens:        total.OutputTokens,
			CacheCreationTokens: total.CacheCreationTokens,
			CacheReadTokens:     total.CacheReadTokens,
			Cost:                total.Cost,
		})
	}
	return out
}

func branchTotalsFromService(in []service.BranchTotal) []BranchTotal {
	out := make([]BranchTotal, 0, len(in))
	for _, total := range in {
		out = append(out, BranchTotal{
			Project:             total.Project,
			Branch:              total.Branch,
			InputTokens:         total.InputTokens,
			OutputTokens:        total.OutputTokens,
			CacheCreationTokens: total.CacheCreationTokens,
			CacheReadTokens:     total.CacheReadTokens,
			Cost:                total.Cost,
		})
	}
	return out
}

func cacheStatsFromService(in service.CacheStats) CacheStats {
	return CacheStats{
		CacheReadTokens:     in.CacheReadTokens,
		CacheCreationTokens: in.CacheCreationTokens,
		UncachedInputTokens: in.UncachedInputTokens,
		OutputTokens:        in.OutputTokens,
		HitRate:             in.HitRate,
		SavingsVsUncached:   in.SavingsVsUncached,
	}
}

File: frontend/src/lib/api/types/usage.ts
/** Usage types — match Go structs in internal/server/usage.go
 *  and internal/db/usage.go */

export interface UsageTotals {
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  totalCost: number;
  copilotAICredits?: number;
}

export interface ModelBreakdown {
  modelName: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface ProjectBreakdown {
  project: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface AgentBreakdown {
  agent: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface BranchBreakdown {
  project: string;
  branch: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface DailyUsageEntry {
  date: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  totalCost: number;
  modelsUsed: string[];
  modelBreakdowns?: ModelBreakdown[];
  projectBreakdowns?: ProjectBreakdown[];
  agentBreakdowns?: AgentBreakdown[];
  branchBreakdowns?: BranchBreakdown[];
}

export interface ProjectTotal {
  project: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface ModelTotal {
  model: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface AgentTotal {
  agent: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface BranchTotal {
  project: string;
  branch: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface CacheStats {
  cacheReadTokens: number;
  cacheCreationTokens: number;
  uncachedInputTokens: number;
  outputTokens: number;
  hitRate: number;
  savingsVsUncached: number;
}

export interface UsageSessionCounts {
  total: number;
  byProject: Record<string, number>;
  byAgent: Record<string, number>;
}

export interface UsageComparison {
  priorFrom: string;
  priorTo: string;
  priorTotalCost: number;
  deltaPct: number;
}

export interface UnsupportedUsage {
  kind: string;
}

export type UsagePairwiseDimension = "model" | "project";

export interface UsagePairwiseComparisonSide {
  totalCost: number;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  totalTokens: number;
  sessionCount: number;
  costPerSession?: number;
  tokensPerSession?: number;
}

export interface UsagePairwiseComparisonDelta {
  totalCostDelta: number;
  totalCostDeltaRatio: number | null;
  inputTokensDelta: number;
  inputTokensDeltaRatio: number | null;
  outputTokensDelta: number;
  outputTokensDeltaRatio: number | null;
  cacheCreationDelta: number;
  cacheCreationDeltaRatio: number | null;
  cacheReadDelta: number;
  cacheReadDeltaRatio: number | null;
  totalTokensDelta: number;
  totalTokensDeltaRatio: number | null;
  sessionCountDelta: number;
  sessionCountDeltaRatio: number | null;
  costPerSessionDelta: number | null;
  costPerSessionRatio: number | null;
  tokensPerSessionDelta: number | null;
  tokensPerSessionRatio: number | null;
}

export interface UsagePairwiseComparisonResponse {
  left: UsagePairwiseComparisonSide;
  right: UsagePairwiseComparisonSide;
  deltas: UsagePairwiseComparisonDelta;
}

export interface UsageSummaryResponse {
  from: string;
  to: string;
  totals: UsageTotals;
  daily: DailyUsageEntry[];
  projectTotals: ProjectTotal[];
  modelTotals: ModelTotal[];
  agentTotals: AgentTotal[];
  branchTotals: BranchTotal[];
  sessionCounts: UsageSessionCounts;
  cacheStats: CacheStats;
  unsupportedUsage?: UnsupportedUsage;
  comparison?: UsageComparison;
}

export interface TopSessionEntry {
  sessionId: string;
  displayName: string;
  agent: string;
  project: string;
  startedAt: string;
  totalTokens: number;
  cost: number;
}

export type TopUsageSessionsResponse = TopSessionEntry[];

export interface UsageParams {
  from?: string;
  to?: string;
  project?: string;
  machine?: string;
  agent?: string;
  model?: string;
  exclude_project?: string;
  exclude_agent?: string;
  exclude_model?: string;
  min_user_messages?: number;
  include_one_shot?: boolean;
  include_automated?: boolean;
  active_since?: string;
  timezone?: string;
}

export interface UsageTopSessionsParams extends UsageParams {
  limit?: number;
}
